HTTP/3 with IPv6

When trying out HTTP/3 support, I noticed on a typical macOS laptop, that it seems IPv6 is used even when configuring the client to IPv4. Following the advice in #584, to force IPv4

```rust
builder = builder
      .http3_prior_knowledge()
      .local_address(IpAddr::from([0, 0, 0, 0]));
```

It seems to still try IPv6 and fail hard. Presumably it should be using IPv4 from the beginning.

```
E       RuntimeError: Request failed: error sending request for url (https://localhost:8081/echo): error sending request: invalid remote address: [::1]:8081
```

Besides that, if not forcing IPv4 like that, when issuing a request to `localhost`, it will first try IPv6 before IPv4. This should generally be fine but AFAICT, on macOS IPv6 can't be used with quic - when trying with `curl --http3`, I get the below output which confirms it tries with IPv6 and quickly falls back to IPv4. This is not noticable. However, when using reqwest, it seems to finally succeed after 30s - I'm guessing there is a 30s timeout on trying to connect with IPv6 before falling back to IPv4 and succeeding. If I change the URL to 127.0.0.1, no issues - I guess the IPv6 -> v4 fallback has to be configured somehow to fail faster.

```
* Host localhost:8081 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:8081...
* SSL Trust: peer verification disabled
* QUIC: connection to ::1 port 8081 refused
* QUIC connect to ::1 port 8081 failed: Could not connect to server
*   Trying 127.0.0.1:8081...
```

--- a/src/async_impl/h3_client/connect.rs
+++ b/src/async_impl/h3_client/connect.rs
@@ -135,10 +135,11 @@ impl H3Connector {
         }
 
         if ipv6_addrs.is_empty() {
-            return Self::try_addresses_static(&self.endpoint,
+            return Self::try_addresses_static(
+                &self.endpoint,
                 &ipv4_addrs,
                 server_name,
-                &self.client_config
+                &self.client_config,
             )
             .await;
         }
@@ -147,7 +148,7 @@ impl H3Connector {
                 &self.endpoint,
                 &ipv6_addrs,
                 server_name,
-                &self.client_config
+                &self.client_config,
             )
             .await;
         }
@@ -155,12 +156,12 @@ impl H3Connector {
         let endpoint = self.endpoint.clone();
         let client_config = self.client_config.clone();
 
-        match Self::try_addresses_static(&endpoint, &ipv6_addrs, server_name, &client_config).await 
+        match Self::try_addresses_static(&endpoint, &ipv6_addrs, server_name, &client_config).await
         {
             Ok(conn) => Ok(conn),
             Err(_) => {
                 Self::try_addresses_static(&endpoint, &ipv4_addrs, server_name, &client_config)
-                .await
+                    .await
             }
         }
     }

--- a/src/async_impl/h3_client/connect.rs
+++ b/src/async_impl/h3_client/connect.rs
@@ -59,6 +59,7 @@ pub(crate) struct H3Connector {
     resolver: DynResolver,
     endpoint: Endpoint,
     client_config: H3ClientConfig,
+    local_addr: Option<IpAddr>,
 }
 
 impl H3Connector {
@@ -86,6 +87,7 @@ impl H3Connector {
             resolver,
             endpoint,
             client_config,
+            local_addr,
         })
     }
 
@@ -117,28 +119,74 @@ impl H3Connector {
         addrs: Vec<SocketAddr>,
         server_name: &str,
     ) -> Result<H3Connection, BoxError> {
-        let mut err = None;
-        for addr in addrs {
-            match self.endpoint.connect(addr, server_name)?.await {
-                Ok(new_conn) => {
-                    let quinn_conn = Connection::new(new_conn);
-                    let mut h3_client_builder = h3::client::builder();
-                    if let Some(max_field_section_size) = self.client_config.max_field_section_size
-                    {
-                        h3_client_builder.max_field_section_size(max_field_section_size);
-                    }
-                    if let Some(send_grease) = self.client_config.send_grease {
-                        h3_client_builder.send_grease(send_grease);
-                    }
-                    return Ok(h3_client_builder.build(quinn_conn).await?);
-                }
-                Err(e) => err = Some(e),
+        if addrs.is_empty() {
+            return Err("no addresses to connect to".into());
+        }
+
+        let (mut ipv6_addrs, mut ipv4_addrs): (Vec<SocketAddr>, Vec<SocketAddr>) =
+            addrs.into_iter().partition(|addr| addr.is_ipv6());
+
+        if let Some(local_ip) = self.local_addr {
+            if local_ip.is_ipv6() {
+                ipv4_addrs.clear();
+            } else {
+                ipv6_addrs.clear();
             }
         }
 
-        match err {
-            Some(e) => Err(Box::new(e) as BoxError),
-            None => Err("failed to establish connection for HTTP/3 request".into()),
+        if ipv6_addrs.is_empty() {
+            return Self::try_addresses_static(&self.endpoint, &ipv4_addrs, server_name, &self.client_config).await;
         }
+        if ipv4_addrs.is_empty() {
+            return Self::try_addresses_static(&self.endpoint, &ipv6_addrs, server_name, &self.client_config).await;
+        }
+
+        let endpoint = self.endpoint.clone();
+        let client_config = self.client_config.clone();
+
+        match Self::try_addresses_static(&endpoint, &ipv6_addrs, server_name, &client_config).await {
+            Ok(conn) => Ok(conn),
+            Err(_) => {
+                Self::try_addresses_static(&endpoint, &ipv4_addrs, server_name, &client_config).await
+            }
+        }
+    }
+
+    async fn try_addresses_static(
+        endpoint: &Endpoint,
+        addrs: &[SocketAddr],
+        server_name: &str,
+        client_config: &H3ClientConfig,
+    ) -> Result<H3Connection, BoxError> {
+        let mut last_err: Option<BoxError> = None;
+
+        for addr in addrs {
+            match endpoint.connect(*addr, server_name) {
+                Ok(connecting) => {
+                    match connecting.await {
+                        Ok(new_conn) => {
+                            let quinn_conn = Connection::new(new_conn);
+                            let mut h3_client_builder = h3::client::builder();
+                            if let Some(max_field_section_size) = client_config.max_field_section_size
+                            {
+                                h3_client_builder.max_field_section_size(max_field_section_size);
+                            }
+                            if let Some(send_grease) = client_config.send_grease {
+                                h3_client_builder.send_grease(send_grease);
+                            }
+                            return Ok(h3_client_builder.build(quinn_conn).await?);
+                        }
+                        Err(e) => {
+                            last_err = Some(Box::new(e) as BoxError);
+                        }
+                    }
+                }
+                Err(e) => {
+                    last_err = Some(Box::new(e) as BoxError);
+                }
+            }
+        }
+
+        Err(last_err.unwrap_or_else(|| "no addresses available".into()))
     }
 }

--- a/src/async_impl/h3_client/connect.rs
+++ b/src/async_impl/h3_client/connect.rs
@@ -59,7 +59,7 @@ pub(crate) struct H3Connector {
     resolver: DynResolver,
     endpoint: Endpoint,
     client_config: H3ClientConfig,
-    local_addr: Option<IpAddr>,
+    bound_addr: IpAddr,
 }
 
 impl H3Connector {
@@ -75,19 +75,23 @@ impl H3Connector {
         // FIXME: Replace this when there is a setter.
         config.transport_config(Arc::new(transport_config));
 
+        // Pipe the local address through to the endpoint creation
         let socket_addr = match local_addr {
             Some(ip) => SocketAddr::new(ip, 0),
-            None => "[::]:0".parse::<SocketAddr>().unwrap(),
+            None => "0.0.0.0:0".parse::<SocketAddr>().unwrap(),
         };
 
         let mut endpoint = Endpoint::client(socket_addr)?;
         endpoint.set_default_client_config(config);
 
+        // Get the actual bound address from the endpoint
+        let bound_addr = endpoint.local_addr()?.ip();
+
         Ok(Self {
             resolver,
             endpoint,
             client_config,
-            local_addr,
+            bound_addr,
         })
     }
 
@@ -126,12 +130,10 @@ impl H3Connector {
         let (mut ipv6_addrs, mut ipv4_addrs): (Vec<SocketAddr>, Vec<SocketAddr>) =
             addrs.into_iter().partition(|addr| addr.is_ipv6());
 
-        if let Some(local_ip) = self.local_addr {
-            if local_ip.is_ipv6() {
-                ipv4_addrs.clear();
-            } else {
-                ipv6_addrs.clear();
-            }
+        if self.bound_addr.is_ipv6() {
+            ipv4_addrs.clear();
+        } else {
+            ipv6_addrs.clear();
         }
 
         if ipv6_addrs.is_empty() {

--- a/src/async_impl/h3_client/connect.rs
+++ b/src/async_impl/h3_client/connect.rs
@@ -78,7 +78,7 @@ impl H3Connector {
         // Pipe the local address through to the endpoint creation
         let socket_addr = match local_addr {
             Some(ip) => SocketAddr::new(ip, 0),
-            None => "0.0.0.0:0".parse::<SocketAddr>().unwrap(),
+            None => "[::]:0".parse::<SocketAddr>().unwrap(),
         };
 
         let mut endpoint = Endpoint::client(socket_addr)?;

--- a/src/async_impl/h3_client/connect.rs
+++ b/src/async_impl/h3_client/connect.rs
@@ -135,19 +135,32 @@ impl H3Connector {
         }
 
         if ipv6_addrs.is_empty() {
-            return Self::try_addresses_static(&self.endpoint, &ipv4_addrs, server_name, &self.client_config).await;
+            return Self::try_addresses_static(&self.endpoint,
+                &ipv4_addrs,
+                server_name,
+                &self.client_config
+            )
+            .await;
         }
         if ipv4_addrs.is_empty() {
-            return Self::try_addresses_static(&self.endpoint, &ipv6_addrs, server_name, &self.client_config).await;
+            return Self::try_addresses_static(
+                &self.endpoint,
+                &ipv6_addrs,
+                server_name,
+                &self.client_config
+            )
+            .await;
         }
 
         let endpoint = self.endpoint.clone();
         let client_config = self.client_config.clone();
 
-        match Self::try_addresses_static(&endpoint, &ipv6_addrs, server_name, &client_config).await {
+        match Self::try_addresses_static(&endpoint, &ipv6_addrs, server_name, &client_config).await 
+        {
             Ok(conn) => Ok(conn),
             Err(_) => {
-                Self::try_addresses_static(&endpoint, &ipv4_addrs, server_name, &client_config).await
+                Self::try_addresses_static(&endpoint, &ipv4_addrs, server_name, &client_config)
+                .await
             }
         }
     }
@@ -162,25 +175,22 @@ impl H3Connector {
 
         for addr in addrs {
             match endpoint.connect(*addr, server_name) {
-                Ok(connecting) => {
-                    match connecting.await {
-                        Ok(new_conn) => {
-                            let quinn_conn = Connection::new(new_conn);
-                            let mut h3_client_builder = h3::client::builder();
-                            if let Some(max_field_section_size) = client_config.max_field_section_size
-                            {
-                                h3_client_builder.max_field_section_size(max_field_section_size);
-                            }
-                            if let Some(send_grease) = client_config.send_grease {
-                                h3_client_builder.send_grease(send_grease);
-                            }
-                            return Ok(h3_client_builder.build(quinn_conn).await?);
+                Ok(connecting) => match connecting.await {
+                    Ok(new_conn) => {
+                        let quinn_conn = Connection::new(new_conn);
+                        let mut h3_client_builder = h3::client::builder();
+                        if let Some(max_field_section_size) = client_config.max_field_section_size {
+                            h3_client_builder.max_field_section_size(max_field_section_size);
                         }
-                        Err(e) => {
-                            last_err = Some(Box::new(e) as BoxError);
+                        if let Some(send_grease) = client_config.send_grease {
+                            h3_client_builder.send_grease(send_grease);
                         }
+                        return Ok(h3_client_builder.build(quinn_conn).await?);
                     }
-                }
+                    Err(e) => {
+                        last_err = Some(Box::new(e) as BoxError);
+                    }
+                },
                 Err(e) => {
                     last_err = Some(Box::new(e) as BoxError);
                 }

--- a/src/async_impl/h3_client/pool.rs
+++ b/src/async_impl/h3_client/pool.rs
@@ -364,7 +364,22 @@ where
 pub(crate) fn extract_domain(uri: &mut Uri) -> Result<Key, Error> {
     let uri_clone = uri.clone();
     match (uri_clone.scheme(), uri_clone.authority()) {
-        (Some(scheme), Some(auth)) => Ok((scheme.clone(), auth.clone())),
+        (Some(scheme), Some(auth)) => {
+            let scheme_str = scheme.as_str();
+            if scheme_str != "https" && scheme_str != "h3" {
+                return Err(Error::new(
+                    Kind::Request,
+                    Some(Box::new(std::io::Error::new(
+                        std::io::ErrorKind::InvalidInput,
+                        format!(
+                            "HTTP/3 only supports 'https' or 'h3' schemes, got: {}",
+                            scheme_str
+                        ),
+                    ))),
+                ));
+            }
+            Ok((scheme.clone(), auth.clone()))
+        }
         _ => Err(Error::new(Kind::Request, None::<Error>)),
     }
 }

